ramp4-pcar4/ramp4-pcar4
# add support for customized grid buttons

### Related Item(s)
No RAMP4 issue exists at the moment. 

This feature is pretty urgently needed for Cumulative Effects. I wrote this up in one day so I'm looking for as much feedback/suggestions as possible!

### Changes
Added a new grid config option: `controls`, which lets you customize which buttons appear in the rows of the grid. The default for this setting is ["zoom", "details"], which will display the two current buttons. 

These new custom buttons are meant to be event-driven. When you create a button, you provide an event name that is emitted when the button is clicked (event handlers should be placed on the host page).

To add a custom button to the grid, add the following object to the `controls` array:
```
{
    actionName: string; // the name of the event to call when the button is pressed
    icon: string; // the icon to display in the button. It can be an svg, an emoji, or text.
    tooltip: string;
}
```

### Notes
GIF demonstrating a grid with two custom buttons. One button opens the settings menu for the layer, and the other combines the actions of the two default buttons (zooms and displays details panel).
![](https://i.imgur.com/aKhNn3g.gif)

### Testing
Steps:
1. Open the [new sample page](https://ramp4-pcar4.github.io/ramp4-pcar4/ce-custom-buttons/demos/index-samples.html?sample=43) (sample 43)
2. Ensure the new custom buttons in the `Releases of Mercury` grid work correctly
3. Ensure only the "zoom" button displays in the `WFSLayer` grid
4. Test other samples and ensure everything still works as expected


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/1957)
